docs/proposals/templates.md

```diff
@@ -0,0 +1,349 @@
+                  - 
+                    name: "MONGODB_USER"
+                    value: "${MONGODB_USER}"
+                  - 
+                    name: "MONGODB_PASSWORD"
+                    value: "${MONGODB_PASSWORD}"
+                  - 
+                    name: "MONGODB_DATABASE"
+                    value: "${MONGODB_DATABASE}"
+```
+
+## Template Expansion API
+
+A server side endpoint for expanding templates is proposed.  The endpoint will inspect the `templateKind` for the template object
+being posted to the endpoint, and then use that information to determine the appropriate template expansion plugin to invoke.
+
+A template expansion plugin is expected to return a `List` of valid k8s objects, which the template processing endpoint
+will then return to the original caller.  This represents the primary contract for templates:  A template api object 
+envelope containing arbitrary content is converted into a list of k8s objects.
+
+The actual mechanism for looking up and invoking a template expansion plugin (as well as for how one is provided to k8s) is
+still being defined.  One suggestion was to essentially run a template expansion plugin as a microserver on the k8s cluster
+itself, possibly within a namespace with a name that matched the `templateKind`.  However dictating specific namespaces
+be available within the cluster for a plugin feels overly burdensome.  A map between the `templateKind` and
+the api endpoint capable of expanding it is necessary, however.
+
+Template processors could also be provided in client form to help template authors perform local testing+debugging.
+
+Nested templates should be handled by the primary endpoint, retrieving and handing each nested template off to the appropriate
+expansion plugin, and then assembling the aggregate list of objects from the top level template and any nested templates.
+
+## Open discussion points
+### Where to define parameters (closed?)
+
+There has been some discussion around where to define parameters that are being injected into a Template
+
+1. In a separate standalone file
+2. Within the Template itself
+
+This proposal suggests including the parameter definitions within the Template, which provides a self-contained structure that
+can be easily versioned, transported, and instantiated without risk of mismatching content.  In addition, a Template can easily
+be validated to confirm that all parameter references are resolveable.
+
+Separating the parameter definitions makes for a more complex process with respect to
+* Editing a template (if/when first class editing tools are created)
+* Storing/retrieving template objects with a central store
+
+### How to define parameters (closed?)
+
+This question should now be closed as the problem is delegated to individual template expansion plugins.  They will
+be provided with a map of parameter values and metadata and can inject/substitute those parameters into the 
+template body based on the scheme they define.
```

Comment:
Open questions:

What types can the parameters be?  Strings of course, numbers & bools maybe? What about lists and finally dicts?  I would personally be happy with just strings, and requiring the template language to convert to e.g. numbers if necessary.  For something like openshift templates you could always use ${foo:int} or something like that.  However if we want to support arbitrary primitives or even compound values I would be OK with that too.

Autogeneration of params for passwords?  Config should be hermetic, in which case you cannot generate random numbers on the inside (without passing in a seed).  So I think there is a case for this and it would be easy to implement.

Should there be any verification at the envelope level?  Obviously if we have types (int string etc) then we have some basic verification.  But do we want to allow specifying only 0 <= i < 65536, or valid URLs, or emails / ip addresses etc.  If we want to drive GUIs, we'll need some hints like should it be a checkbox or a slider, etc., which ties into this.